<a href="https://colab.research.google.com/github/SelvamRaju/Natural-Language-Processing-Statistical-NLP/blob/main/Statistical_NLP_Part_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Domain : Digital Content Management

In [2]:
# Imports
import pandas as pd
import numpy as np
import os

import warnings
warnings.filterwarnings("ignore")

1 . Read and Analyse Dataset.

  A. Clearly write outcome of data analysis(Minimum 2 points)

  B. Clean the Structured Data

  i . Missing value analysis and imputation.

  ii . Eliminate Non-English textual data.
  
  Hint : Refer ‘langdetect’ library to detect language of the input text)

In [3]:
# Drive mounting
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
#Declaring the working directory (Directory in Drive)
os.chdir('/content/drive/MyDrive/Colab Notebooks/Data/Statistical-NLP/');

In [5]:
#Listing the files under the current working directory
!ls

blogtext.csv


In [6]:
#Importing CSV dataset
df = pd.read_csv('blogtext.csv')
df.head()

id  ...                                               text
0  2059027  ...             Info has been found (+/- 100 pages,...
1  2059027  ...             These are the team members:   Drewe...
2  2059027  ...             In het kader van kernfusie op aarde...
3  2059027  ...                   testing!!!  testing!!!          
4  3581210  ...               Thanks to Yahoo!'s Toolbar I can ...

[5 rows x 7 columns]

In [7]:
# Shape of dataset
df.shape

(681284, 7)

In [8]:
# Getting unique topics
df['topic'].unique()

array(['Student', 'InvestmentBanking', 'indUnk', 'Non-Profit', 'Banking',
       'Education', 'Engineering', 'Science', 'Communications-Media',
       'BusinessServices', 'Sports-Recreation', 'Arts', 'Internet',
       'Museums-Libraries', 'Accounting', 'Technology', 'Law',
       'Consulting', 'Automotive', 'Religion', 'Fashion', 'Publishing',
       'Marketing', 'LawEnforcement-Security', 'HumanResources',
       'Telecommunications', 'Military', 'Government', 'Transportation',
       'Architecture', 'Advertising', 'Agriculture', 'Biotech',
       'RealEstate', 'Manufacturing', 'Construction', 'Chemicals',
       'Maritime', 'Tourism', 'Environment'], dtype=object)

In [9]:
# Count of unique topics
len(df['topic'].unique())

40

In [10]:
# Count from groupby topic
df.groupby(['topic']).count()

id  gender     age    sign    date    text
topic                                                                  
Accounting                 3832    3832    3832    3832    3832    3832
Advertising                4676    4676    4676    4676    4676    4676
Agriculture                1235    1235    1235    1235    1235    1235
Architecture               1638    1638    1638    1638    1638    1638
Arts                      32449   32449   32449   32449   32449   32449
Automotive                 1244    1244    1244    1244    1244    1244
Banking                    4049    4049    4049    4049    4049    4049
Biotech                    2234    2234    2234    2234    2234    2234
BusinessServices           4500    4500    4500    4500    4500    4500
Chemicals                  3928    3928    3928    3928    3928    3928
Communications-Media      20140   20140   20140   20140   20140   20140
Construction               1093    1093    1093    1093    1093    1093
Consulting                 5862    5862    5862    5862    5862    5862
Education                 29633   29633   29633   29633   29633   29633
Engineering               11653   11653   11653   11653   11653   11653
Environment                 592     592     592     592     592     592
Fashion                    4851    4851    4851    4851    4851    4851
Government                 6907    6907    6907    6907    6907    6907
HumanResources             3010    3010    3010    3010    3010    3010
Internet                  16006   16006   16006   16006   16006   16006
InvestmentBanking          1292    1292    1292    1292    1292    1292
Law                        9040    9040    9040    9040    9040    9040
LawEnforcement-Security    1878    1878    1878    1878    1878    1878
Manufacturing              2272    2272    2272    2272    2272    2272
Maritime                    280     280     280     280     280     280
Marketing                  4769    4769    4769    4769    4769    4769
Military                   3128    3128    3128    3128    3128    3128
Museums-Libraries          3096    3096    3096    3096    3096    3096
Non-Profit                14700   14700   14700   14700   14700   14700
Publishing                 7753    7753    7753    7753    7753    7753
RealEstate                 2870    2870    2870    2870    2870    2870
Religion                   5235    5235    5235    5235    5235    5235
Science                    7269    7269    7269    7269    7269    7269
Sports-Recreation          3038    3038    3038    3038    3038    3038
Student                  153903  153903  153903  153903  153903  153903
Technology                42055   42055   42055   42055   42055   42055
Telecommunications         3891    3891    3891    3891    3891    3891
Tourism                    1942    1942    1942    1942    1942    1942
Transportation             2326    2326    2326    2326    2326    2326
indUnk                   251015  251015  251015  251015  251015  251015

In [11]:
#info about dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 681284 entries, 0 to 681283
Data columns (total 7 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   id      681284 non-null  int64 
 1   gender  681284 non-null  object
 2   age     681284 non-null  int64 
 3   topic   681284 non-null  object
 4   sign    681284 non-null  object
 5   date    681284 non-null  object
 6   text    681284 non-null  object
dtypes: int64(2), object(5)
memory usage: 36.4+ MB


In [12]:
#null check
df.isnull().any()

id        False
gender    False
age       False
topic     False
sign      False
date      False
text      False
dtype: bool

In [13]:
# Displaying the datatypes
df.dtypes.to_frame(name = 'Data Type')

Data Type
id         int64
gender    object
age        int64
topic     object
sign      object
date      object
text      object

**Data Analysis outcome:**


*   *The dataset has 681284 rows of data with 7 variable/attributes of information*
*   *There are 40 unique topics where all the ~681K row fall under.*
*   *id and age are the 2 numerice variables*


No missing values identified in the analysis

In [14]:
# Cropping the dataset for computation. Taking the 3K into account for trainning

df = df[:3000]
df.shape

(3000, 7)

In [15]:
# Eliminating non english textual data using langdetect library

!pip install langdetect
from langdetect import detect

def detect_english(text):
    try:
        return detect(text) == 'en'
    except:
        return False

     |████████████████████████████████| 981 kB 5.2 MB/s 
  Created wheel for langdetect: filename=langdetect-1.0.9-py3-none-any.whl size=993242 sha256=af5b3315a70dae62652c265c8d9e7a60993435d4c033c81e993ad876397a8249
  Stored in directory: /root/.cache/pip/wheels/c5/96/8a/f90c59ed25d75e50a8c10a1b1c2d4c402e4dacfa87f3aff36a
Successfully built langdetect


In [16]:
df = df[df['text'].apply(detect_english)]
df.shape

(2821, 7)

**Langdetect has eliminated some of the non-english textual data and out of 5000 we now have 4721 english data for computation.**

2 . Preprocess unstructured data to make it consumable for model training.

  A. Eliminate All special Characters and Numbers

  B. Lowercase all textual data

  C. Remove all Stopwords

  D. Remove all extra white spaces


In [17]:
# Eliminating stop words using NLTK and removing white spaces and converting texts to lower case

import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
import string, re

stop_words = set(stopwords.words('english'))

# Printing all stop words
print('Number of stop words : ', len(stop_words))

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
Number of stop words :  179


In [18]:
#Removing Special characters and numbers
df.text = df.text.apply(lambda x: re.sub('[^A-Za-z]+', ' ', x))

# lowercase conversion
df.text = df.text.apply(lambda x: x.lower())

# removing spaces
df.text = df.text.apply(lambda x: x.strip())

# #Removing stopwords
from nltk.corpus import stopwords
stopwords = set(stopwords.words('english'))
df.text = df.text.apply(lambda x: ' '.join([word for word in x.split() if word not in stopwords]))

In [19]:
df.head(5)

id  ...                                               text
0  2059027  ...  info found pages mb pdf files wait untill team...
1  2059027  ...  team members drewes van der laag urllink mail ...
2  2059027  ...  het kader van kernfusie op aarde maak je eigen...
3  2059027  ...                                    testing testing
4  3581210  ...  thanks yahoo toolbar capture urls popups means...

[5 rows x 7 columns]

3 . Build a base Classification model

  A. Create dependent and independent variables
    
  Hint: Treat ‘topic’ as a Target variable.

  B. Split data into train and test.

  C. Vectorize data using any one vectorizer.

  D. Build a base model for Supervised Learning - Classification.

  E. Clearly print Performance Metrics.
    
  Hint: Accuracy, Precision, Recall, ROC-AUC

In [20]:
# Merging gender, age,topic and sign variables to make a grouped label for multilabel classification
#cropped_df['labels'] = cropped_df['gender'] +',' + cropped_df['age'].astype(str) + ',' + cropped_df['topic'] + ',' + cropped_df['sign']

df['labels'] = df.apply(lambda row: [row['gender'], str(row['age']), row['topic'], row['sign']], axis=1)

df['labels'].head()

0                   [male, 15, Student, Leo]
1                   [male, 15, Student, Leo]
2                   [male, 15, Student, Leo]
3                   [male, 15, Student, Leo]
4    [male, 33, InvestmentBanking, Aquarius]
Name: labels, dtype: object

In [21]:
df = df[['text', 'labels']]
df.head()

text                                   labels
0  info found pages mb pdf files wait untill team...                 [male, 15, Student, Leo]
1  team members drewes van der laag urllink mail ...                 [male, 15, Student, Leo]
2  het kader van kernfusie op aarde maak je eigen...                 [male, 15, Student, Leo]
3                                    testing testing                 [male, 15, Student, Leo]
4  thanks yahoo toolbar capture urls popups means...  [male, 33, InvestmentBanking, Aquarius]

In [22]:
# Train Test split
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df.text.values,df.labels.values,test_size=0.20, random_state=42)


In [23]:
# Vectorizing using Count Vectorizer
from sklearn.feature_extraction.text import CountVectorizer

count_vectorizer = CountVectorizer(binary=True, ngram_range=(1, 2))

X_train_BOW = count_vectorizer.fit_transform(X_train)
X_test_BOW = count_vectorizer.transform(X_test)

In [24]:
# Printing the length of vocabulary
len(count_vectorizer.vocabulary_)

176360

In [25]:
count_vectorizer.get_feature_names()[:10]

['aa',
 'aa anger',
 'aa compared',
 'aa nice',
 'aaa',
 'aaa take',
 'aaa travel',
 'aaaaaah',
 'aaaah',
 'aaaah wisdom']

In [26]:
# term-document matrix
X_train_BOW.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [27]:
label_counts=dict()

for labels in df.labels.values:
    for label in labels:
        if label in label_counts:
            label_counts[str(label)]+=1
        else:
            label_counts[str(label)]=1

In [28]:
# Printing the dictionary
label_counts

{'14': 73,
 '15': 285,
 '16': 23,
 '17': 147,
 '23': 84,
 '24': 319,
 '25': 106,
 '26': 42,
 '27': 86,
 '33': 94,
 '34': 6,
 '35': 1476,
 '37': 19,
 '39': 30,
 '41': 14,
 '44': 3,
 '45': 14,
 'Accounting': 2,
 'Aquarius': 273,
 'Aries': 1568,
 'Arts': 2,
 'Banking': 16,
 'BusinessServices': 9,
 'Cancer': 63,
 'Capricorn': 77,
 'Communications-Media': 14,
 'Education': 116,
 'Engineering': 119,
 'Gemini': 21,
 'Internet': 20,
 'InvestmentBanking': 70,
 'Leo': 54,
 'Libra': 312,
 'Museums-Libraries': 2,
 'Non-Profit': 45,
 'Pisces': 2,
 'Sagittarius': 103,
 'Science': 31,
 'Scorpio': 240,
 'Sports-Recreation': 75,
 'Student': 387,
 'Taurus': 71,
 'Technology': 1476,
 'Virgo': 37,
 'female': 706,
 'indUnk': 437,
 'male': 2115}

4 . Improve Performance of model.

  A. Experiment with other vectorisers.

  B. Build classifier Models using other algorithms than base model.

  C. Tune Parameters/Hyperparameters of the model/s.

  D. Clearly print Performance Metrics.

  Hint: Accuracy, Precision, Recall, ROC-AUC

In [29]:
# Using multi label binarizer

from sklearn.preprocessing import MultiLabelBinarizer

multiLabelBinarizer = MultiLabelBinarizer(classes=sorted(label_counts.keys()))
y_train = multiLabelBinarizer.fit_transform(y_train)
y_test = multiLabelBinarizer.transform(y_test)

In [30]:
from sklearn.multiclass import OneVsRestClassifier
from sklearn.linear_model import LogisticRegression

classifier = LogisticRegression(solver='lbfgs')
classifier = OneVsRestClassifier(classifier)

In [31]:
#Fitting the classifier

classifier.fit(X_train_BOW, y_train)

OneVsRestClassifier(estimator=LogisticRegression())

In [32]:
#Get predicted values
predicted_labels = classifier.predict(X_test_BOW)
predicted_scores = classifier.decision_function(X_test_BOW)

In [33]:
# Get inverse transform values
pred_inversed = multiLabelBinarizer.inverse_transform(predicted_labels)
y_test_inversed = multiLabelBinarizer.inverse_transform(y_test)

In [37]:
# Printing the actual and predicted values
for i in range(10):
  print('Text : ', X_test[i])
  print('Actual    : ', y_test_inversed[i])
  print('Predicted : ', pred_inversed[i])
  print('--------------------------')

Text :  trying watch world news stuff okay see little iraqi palestinian kids throwing stuff boys say hell parents mean seriously dude lived mom dad would like hel l oh get rear gear get house usa marching around mr roger neighborhood mk gernades safe play street times like want move cocoa beach want surf miss makes sad today good day reason depressed maybe thinking everything like see people good ya kno well know yeah seriously hot outside curently house crazy may like would comfortable ds comfortable mother driving crazy like oh hey home none friends felt like caring tonight guess speak words mother never ah repeats head going explode swear source constant headache going impossible stay summer hope knows def going living harry bob dan house care whose long need hobby besides online think spend hours end front thing ah mazing think rambled enough god shut help anyone save tonight owe something big promise
Actual    :  ('15', 'Libra', 'Student', 'female')
Predicted :  ('15', 'Libra', 'S

In [40]:
#Performance metrics -MICRO/MACRO/WEIGHTED/SAMPLES
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score
from sklearn.metrics import average_precision_score
from sklearn.metrics import recall_score

def print_performance_scores(y_val, predicted, averageType):
    print('----------------'+averageType+'--------------------------')
    print('Accuracy score: ', accuracy_score(y_val, predicted))
    print('F1 score: ', f1_score(y_val, predicted, average=averageType))
    print('Average precision score: ', average_precision_score(y_val, predicted, average=averageType))
    print('Average recall score: ', recall_score(y_val, predicted, average=averageType))

In [47]:
print_performance_scores(y_test, predicted_labels, 'micro')
print_performance_scores(y_test, predicted_labels, 'macro')
print_performance_scores(y_test, predicted_labels, 'weighted')
print_performance_scores(y_test, predicted_labels, 'samples')

----------------micro--------------------------
Accuracy score:  0.5097345132743363
F1 score:  0.7188202949262684
Average precision score:  0.556500412969257
Average recall score:  0.636283185840708
----------------macro--------------------------
Accuracy score:  0.5097345132743363
F1 score:  0.25402953495908803
Average precision score:  nan
Average recall score:  0.20387574118628823
----------------weighted--------------------------
Accuracy score:  0.5097345132743363
F1 score:  0.653114818750386
Average precision score:  0.5864082927996913
Average recall score:  0.636283185840708
----------------samples--------------------------
Accuracy score:  0.5097345132743363
F1 score:  0.6754670599803344
Average precision score:  0.647731908617335
Average recall score:  0.636283185840708


Trying other vectorisers

In [49]:
# TF-IDF or Term Frequency–Inverse Document Frequency

from sklearn.feature_extraction.text import TfidfVectorizer


tfidfVectorizer = CountVectorizer(binary=True, ngram_range=(1, 2))

X_train_tfid = tfidfVectorizer.fit_transform(X_train)
X_test_tfid = tfidfVectorizer.transform(X_test)

# using the same classifier and binarizer used for CountVectorizer

classifier.fit(X_train_tfid, y_train)

#Get predicted values
predicted_labels = classifier.predict(X_test_tfid)
predicted_scores = classifier.decision_function(X_test_tfid)

# Get inverse transform values
pred_inversed = multiLabelBinarizer.inverse_transform(predicted_labels)
y_test_inversed = multiLabelBinarizer.inverse_transform(y_test)

# Printing the actual and predicted values
for i in range(10):
  print('Text : ', X_test[i])
  print('Actual    : ', y_test_inversed[i])
  print('Predicted : ', pred_inversed[i])
  print('--------------------------')


Text :  trying watch world news stuff okay see little iraqi palestinian kids throwing stuff boys say hell parents mean seriously dude lived mom dad would like hel l oh get rear gear get house usa marching around mr roger neighborhood mk gernades safe play street times like want move cocoa beach want surf miss makes sad today good day reason depressed maybe thinking everything like see people good ya kno well know yeah seriously hot outside curently house crazy may like would comfortable ds comfortable mother driving crazy like oh hey home none friends felt like caring tonight guess speak words mother never ah repeats head going explode swear source constant headache going impossible stay summer hope knows def going living harry bob dan house care whose long need hobby besides online think spend hours end front thing ah mazing think rambled enough god shut help anyone save tonight owe something big promise
Actual    :  ('15', 'Libra', 'Student', 'female')
Predicted :  ('15', 'Libra', 'S

In [50]:
# Classification metrics for TF-IDF

print_performance_scores(y_test, predicted_labels, 'micro')
print_performance_scores(y_test, predicted_labels, 'macro')
print_performance_scores(y_test, predicted_labels, 'weighted')
print_performance_scores(y_test, predicted_labels, 'samples')

----------------micro--------------------------
Accuracy score:  0.5097345132743363
F1 score:  0.7188202949262684
Average precision score:  0.556500412969257
Average recall score:  0.636283185840708
----------------macro--------------------------
Accuracy score:  0.5097345132743363
F1 score:  0.25402953495908803
Average precision score:  nan
Average recall score:  0.20387574118628823
----------------weighted--------------------------
Accuracy score:  0.5097345132743363
F1 score:  0.653114818750386
Average precision score:  0.5864082927996913
Average recall score:  0.636283185840708
----------------samples--------------------------
Accuracy score:  0.5097345132743363
F1 score:  0.6754670599803344
Average precision score:  0.647731908617335
Average recall score:  0.636283185840708
